proposals/blittable.md

```diff
@@ -0,0 +1,161 @@
+
+## Motivation
+[motivation]: #motivation
+
+The primary motivation is to make it easier to author low level interop code in C#.  Blittable types are one of the core building blocks for interop code, yet abstracting around them is difficult today due to a lack of declarative language support.  
+
+This is the case even though blittable types are well defined in the language spec and many features, like pointers, can operate only on them.  The language chooses to let structs implicitly opt into being blittable by virtue of their construction with no avenue for opting out of such a classification.
+
+While attractive in small applications this is more difficult to manage across a large set of libraries authored by different developers.  It means small field additions to structs can cause compile and runtime breaks in downstream consumers with no warning to the developers that made the change.  This spooky action at a distance is one of the core problems with having an implicit opt in model here. 
+
+A declarative, explicit opt in model makes it easier to code in this area.  Developers can be very explicit about the types they intend for interop.  This gives them compiler help when mistakes are made in their application and in any libraries they consume.  
+
+The lack of compile time support also makes it difficult to abstract over blittable types.  It's not possible for instance to author common helper routines using generic code:
+
+``` c#
+void Hash<T>(T value) where T : blittable struct
+{
+    using (T* p = &value) { 
+        ...
+    }
+}
+```
+
+Instead developers are forced to rewrite virtually the same code for all of their blittable types:
+
+``` c#
+void Hash(Point p) { 
+    ...
+}
+
+void Hash(Time t) { 
+    ...
+}
+```
+
+The lack of constraints here also make it impossible to have efficient conversions between streams of data and more structured types.  This is an operation that is common in networking stacks and serialization layers:
+
+``` c#
+Span<byte> Convert<T>(ref T value) where T : blittable {
+    ...
+}
+```
+
+Such routines are advantageous because they are provably safe at compile time and allocation free.  Interop authors today can not due this (even though it's at a layer where perf is critical).  Instead they need to rely on allocating routines that have expensive runtime checks to verify values are correctly blittable.
+
+## Detailed design
+[design]: #detailed-design
+
+The language will introduce a new declaration modifier named `blittable` that can be applied to `struct` definitions.  
+
+``` c#
+blittable struct Point 
+{
+    public int X;
+    public int Y;
+}
+```
+
+The compiler will enforce that the fields of such a struct definition fit one of the following categories:
+
+- `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, or `bool`.
```

Comment:
I think if we are explicitly stating that this is for 'unmanaged types' (that is types for which the unmanaged and managed representation is bit for bit identical) then we should not be including 'char' or 'bool', since the statement is not always true for these types.

String is technically blittable in some cases (and there is logic for pinning a string rather than copying in those instances).